### DISCUSSION TITLE
Special public servers are not visible in the list.

### DISCUSSION THREAD

**Discussion by Nadolisova** [jamulussoftware/jamulus#2627] [Bug Investigation]
The problem has existed since November 2021. Suddenly servers that were close in location disappeared from the list. When trying to detect them, we contacted the server administrator and tried different options with his help. On the same server he raised private server on port 60606, I could connect it. But the public one was not visible and hung in “trying to connect” when connect by IP:port. 
If I change network to a mobile operator, then the server was visible and available. I used a workaround several times: I connected via a mobile network to the server, and then changed it to a local one and Jamulus reconnected without problems, so we rehearsed. But since Feb 2022, this workaround suddenly stopped working. Changing server and client versions did not help. 
There are no restrictions on the router. I also changed the router and it did not help. Provider has no restrictions too.

The solution was to bring up the server locally, it worked great until now. Today I didn't see him on the list, although it is in the next room within my local network. Available from the mobile network, but not available from the local network, which has no restrictions. After restarting the server, it became visible and available locally. 

It’s not network or port restriction problem. Some problem is hidden in discovering servers in list inside Jamulus software.

	Os tested: Ubuntu 20.04.4 LTS, Windows 11, MacOS BigSur
	Jamulus tested:  3.8.1,  3.8.2. 
	Server ver: 3.8.0dev-379ddf3, 3.8.2

**Comment by gene96817** [comment]
From each server, what is their IP address?
From the client perspective, what IP address is the client using to connect to the server?
What do you mean by "servers that were close in location"?
   -- close as in they are a few meters apart?
   -- close is in that are on the same private network?
   -- something else?

I am expecting an unexpected firewall issue.
It is possible your suspicion there is a bug causing "some problem is hidden in discovering servers in list inside Jamulus software".
If true, then we need to discover what table elements from the two sever table records ("servers that were close in location") is breaking the table logic.

**Reply by pljones** [reply]
It's always worth checking something like this using https://explorer.jamulus.io to see whether it's a local issue or not.  If Jamulus Explorer is having a problem with just your server, it's likely a local issue.

**Reply by Nadolisova** [reply]
> From each server, what is their IP address? 

My server is in Any Genre 2:
MoscowServer	Moscow	Russia	92.38.87.64	(36571) 22124
Server not seeing from november 2021 in Any Genre 1:
Geos24/7	Saint Petersburg	Russia	176.57.220.151	22124

> From the client perspective, what IP address is the client using to connect to the server? 

I always use connect to name or public address in bottom line "Server address" that I see in Jamulus Explorer. And used also local network address for my personal server when tested it issue.

> What do you mean by "servers that were close in location"? -- close as in they are a few meters apart? -- close is in that are on the same private network? -- something else?

Close to me" i mean server is in Russia. There is only one today. 

> I am expecting an unexpected firewall issue. It is possible your suspicion there is a bug causing "some problem is hidden in discovering servers in list inside Jamulus software". If true, then we need to discover what table elements from the two sever table records ("servers that were close in location") is breaking the table logic.

It's not FW issue, because after server restart I see server and can connect to.

**Reply by gene96817** [reply]
well, this is very different than what I inferred from your previous message.
Here is what I now understand.

1- Jamulus Explorer sees both servers.
2- The two Jamulus servers are in Moscow and St. Petersburg
3- Your clients see Jamulus Explorer and the desired Jamulus servers listed.

I don't follow the ISP network issues in Russia. Now I am wondering:
4- Are any "national" routing or traffic/content management policies (i.e. firewall) issues in play here.
     Any reason that Jamulus traffic might fall under some national policy to restrict the traffic?
     Are there any national policies that could inadvertently affect Jamulus?
5- An interesting test would be to install a small webserver on the two Jamulus servers. 
    Can your users access the two webservers? 
5A- there could be something flakey in the routing tables? (mistakes happen, sometimes on purpose, sometimes just mistakes)
5B- if the web connections are reliable but Jamulus is not, there might be some data traffic policy in play.

**Reply by Nadolisova** [reply]
> 1- Jamulus Explorer sees both servers. 2- The two Jamulus servers are in Moscow and St. Petersburg 3- Your clients see Jamulus Explorer and the desired Jamulus servers listed.

Yes. Client sees both servers in Explorer. And doesnt see server in list inside Jamulus. At same time client 

> I don't follow the ISP network issues in Russia. Now I am wondering: 4- Are any "national" routing or traffic/content management policies (i.e. firewall) issues in play here. Any reason that Jamulus traffic might fall under some national policy to restrict the traffic? Are there any national policies that could inadvertently affect Jamulus? 

I dont know any special rules for Russia providers. There was another server in Moscow, it works well, but now we lost unfortunately its admin...

> 5- An interesting test would be to install a small webserver on the two Jamulus servers. Can your users access the two webservers? 

Lets take one of them. Geos,  176.57.220.151. Yes, there is web and its accessible any time no problem.

> 5A- there could be something flakey in the routing tables? (mistakes happen, sometimes on purpose, sometimes just mistakes), 5B- if the web connections are reliable but Jamulus is not, there might be some data traffic policy in play.

As admin said there are no restrictions in policies. And no problem in routing because web is ok at same time.


If it will help I can collect traffic by Wireshark.

**Reply by gene96817** [reply]
It would be helpful to confirm the 5 above for both servers.

If you have reliable web access to both servers but problems with Jamulus, it suggests there is a routing problem with one of the ISPs.
1- it could be a mistake in the routing  table
2- it the problem is intermittent, it could be a problem with a secondary path
3- it could be a problem in how UDP is handled but the TCP works.

None of the clues suggest a Jamulus problem.
Note the test conditions above use traffic between a Russian IP address to a non-Russian IP address.
The problem is between two Russian IP addresses. It could be a routing problem between two Russian ISPs.

**Reply by pljones** [reply]
I get a similar issue with quite a few servers: for example, the Rainbow (Jazz genre) servers in Italy on 93.42.88.44 are visible in Jamulus Explorer and I can ping them - but I can't see them or connect with Jamulus.  Somewhere, routing drops the packets.  It's very odd, I run the Jazz directory, the server registers fine with it and it's generally visible.  Just not for me (from the same machine as the directory or within the LAN - and the addresses are working).

**Reply by gene96817** [reply]
There is clearly a routing or network path problem.
Sorry, I can't read the Jamulus code.
How much UDP packet loss can Jamulus tolerate?
It could be that the Jamulus UDP packets at startup are too big for some of the client's network paths.
I recall seeing some past discussions about avoiding the use of UDP packet fragmentation.
We may be seeing a reason to fragment packets.

**Reply by Nadolisova** [reply]
I can start another instance Jamulus on the same server at another port. It works. 
If i restart only service than no problem - server available. That means no problem in routing and filtering ports.

I think I find path to reproduce error on my server MoscowServer in Jenre 2. Maybe this will help to discover my problem. 
If i restart full server: jamulus-headless.service starts with _**bad address**_ and after 15 minutes it registers in directory but I can't connect it. If i restart service only then I can see and connect. 


Now MoscowServer restarted same way and I cannot see it. My friend from another city cannot see it too.


may 23 10:56:37 server jamulus[701]: Server Registration Status update: **Bad address**
may 23 11:11:37 server jamulus[701]: Server Registration Status update: Registration requested
may 23 11:11:37 server jamulus[701]: Server Registration Status update: Registered



1. What does it mean bad address? Service starts before networking up on server?
2. Why service registers after bad address in directory but not fully?

**Reply by Nadolisova** [reply]
Another investigation.  Server accessible by second port  for example it's now  (51685). I can connect to it! 
And client can't see in list other servers that have second port. 
Jamulus check ping by main port 22124, fails and doesn't show server in client at all?

Why some servers have second port?

**Reply by pljones** [reply]
> Why some servers have second port?

What do you mean?

**Reply by Nadolisova** [reply]
> What do you mean?

In Jamulus explorer my server for example has in Port:

63	MoscowServer	Moscow	Russia	0 / 16	49	92.38.87.64	**(51685) 22124**	3.8.2	Linux

**Reply by pljones** [reply]
22124 is the port number your server is telling the directory it's using.

However, if Jamulus Explorer can't get a response on that port, it goes looking for a server and reports the first port it finds with a server response.  That's the 51685.

You need to set your router firewall / port forwarding / related settings to understand that when you say "22124", you don't want it being changed.

**Reply by Nadolisova** [reply]
That means jamulus can change port on its own?  Where i can fix it?

**Reply by gene96817** [reply]
I encountered this behavior when I was debugging my Jamulus installation. Part of that process, I was doing frequent rebooting. While you can tell Jaumulus to use port 22124, if the server is behind a firewall using NAT, if port 22124 is in use, the UDP packets will be mapped to a free port. In your case, the packets are mapped to 51685. This is the port Jamulus Explorer is seeing in your UDP packets.  In this scenario, 22124 available the first time the Jamulus server starts up. On reboot, the firewall does not know about the reboot and finds a new port. Eventually the firewall will declare 22124 free.

You can fix this by defining a fixed mapping in your router.
-- set DHCP to always assign your servers MAC address the same local IP address.
-- set a fixed mapping for the firewall to always mapping localIP:22124 to the routers publicIP:22124.
-- reboot the router to flush the old states.
The result is Jamulus Explorer will always see UDP packets from your server using port 22124.
Note this only works with one server behind your firewall. You can extend the same method by assigning a fixed port (not 22124) to the second server.

**Reply by pljones** [reply]
> That means jamulus can change port on its own? Where i can fix it?

That's not what I said.  I said your router is changing the port and I said you need to fix your router.

**Reply by Nadolisova** [reply]
Thank you, i think i understand how it works. It expain behavior of my server. But it doesn't explain why server Geos hidden with fixed port.

I'll try some tests.
